### DISCUSSION TITLE
Parametric Remote capture ignoring NEC repeat code?

### DISCUSSION THREAD

**Discussion by eriknz** [bengtmartensson/IrScrutinizer#530] [Decoding issues]
I have an Acer projector remote. It seems like it's using NEC1 and NEC. Most of the buttons when pressed, transmit a command followed by a repeat code (9ms pulse followed by a 2.25ms space and a 560µs pulse). If they're held down, they only transmit the command once then just repeat the repeat code (I guess to prevent sending the command multiple times?).
But there are arrow/cursor keys that don't have the repeat signal and when held down they re-transmit the command, I guess to make the cursor continue to move if you hold the button down.
In the Scrutinize signal tab and the Scrutinize Remote Raw Remote tab, the buttons with the repeat code are detected as NEC1. But in the Parametric Remote tab they're detected as NEC.
For the arrow/cursor buttons without the repeat code, the Scrutinize signal and Scrutinize Remote Raw tabs detect them as NEC but the Parametric Remote tab detects it as NEC2.

**Comment by bengtmartensson** [comment]
As you seem to be aware of, NEC1 consists of one long segment ("payload") and a repeat consisting of just two pulses ("ditto"). In NEC2, the payload is instead repeated. If exactly one payload, and no repeats, are received. this is decoded as "NEC". This is almost always an incompletely captured signal. So I ask you first to make absolutely sure that some keys (which ones?) really do not repeat (press the key for at least a second). You can try with and without the signal cleaner and the repeat finder (found under Options). What exact model of Acer projector do you have?

> For the arrow/cursor buttons without the repeat code, the Scrutinize signal and Scrutinize Remote Raw tabs detect them as NEC but the Parametric Remote tab detects it as NEC2.

_If_ that is true, it is a bug. I tried with two IrScrutinizers, sent a NEC1 with a count of 1 (that is "NEC"), and captured with the second, and I could not reproduce; it was all recognized as "NEC". Please give more details, that I can reproduce.

**Reply by eriknz** [reply]
I can confirm that all the buttons do continue to transmit if held down. However, all except the arrow buttons repeat the 2 pulse repeat code. The arrow buttons repeat the whole command. On the projector, holding down the arrow keys make the selected option continue to scroll. Holding down any of the other keys doesn't seem to have any effect over just pressing them once.  

Hardware I'm using an Arduino Nano with what I think is a demodulating receiver (from this https://wiki.keyestudio.com/Ks0026_keyestudio_Digital_IR_Receiver_Module but removed from the board as I think the capacitors were affecting the signal) and followed your tutorial on setting it up as a Girs Client. I have "Use receiver for capture" selected.

The projector is an Acer H6530BD, the remote says it's model No: J-25041. I've also tested the NEC1 signal stuff with the remote from a Logitech Z607 speaker remote and got the same results.

I found that for super short key presses, all the buttons decoded as NEC regardless of repeat finder settings or which panel I used. I guess they were just too short for the repeat code or a repeat of the command to be sent.

The below notes are for single brief (but not super short) key presses.

In the Scrutinize signal panel, for the arrow keys (signal doesn't have a repeat code), with repeat finder off it gets decoded as NEC2. With repeat finder on it gets decoded as NEC. In the Parametric Remote panel, it is always decoded as NEC2. Looking at the Pronto code for NEC and NEC2, it seems it's the same except for NEC has all the burst pairs in Sequence 1 whereas NEC2 has them all in sequence 2 (the repeat sequence). Since the arrow keys on this remote are repeating, I guess that means they must be NEC2?

But for the standard keys (with the repeat code following the command) in the Scrutinize signal panel, they're decoded as NEC1 regardless of repeat finder setting. However in the Parametric Remote panel, they're decoded as NEC1 with repeat finder off, or NEC with it on.


To summarize:
In the Parametric Remote panel, with Repeat Finder on, NEC1 signals get decoded incorrectly as NEC. With it off they get decoded correctly as NEC1.
However I think the Parametric Remote panel is decoding the NEC2 signals correctly regardless of repeat finder setting. But in the Scrutinize signal panel with Repeat finder on, it is incorrectly decoding them as NEC, ignoring the fact that it's repeating (maybe that's how you want repeat finder to behave?).

Here's a spreadsheet with my test results:
https://docs.google.com/spreadsheets/d/1Qpb15ayo4J40wYuP4h1SSWHAUdfMVR9k8SnJz0hz30c/


I hope this makes sense. I feel like figuring it out is making my brain work a lot harder than it usually has to.

**Comment by bengtmartensson** [comment]
I found and fixed a bug, #531. Thank you for your observations! Please download the [snapshot version](https://github.com/bengtmartensson/IrScrutinizer/releases/tag/ci-build).

Actually, this is the first time I see a device that uses both NEC1 and NEC2. [GlobalCache's Controltower](https://irdb.globalcache.com/) has a file for Acer Projector, and it seems to confir your findings.

> I feel like figuring it out is making my brain work a lot harder than it usually has to.

Great!

Thank you!

**Reply by eriknz** [reply]
I've tested the snapshot version and can confirm now with Repeat Finder on, in the Scrutinize Panel, NEC2 gets decoded correctly. With Repeat Finder off, NEC2 gets decoded as NEC, which I guess makes sense since it's not looking for the repeat (this is however different to the Parametric Remote panel which decodes NEC2 correctly regardless of Repeat Finder setting).
But it seems like NEC1 decoding is broken now. See the tables below:

### Scrutinize signal panel, continuous capture.

| Input  | Repeat Finder off | Repeat Finder on |
|-|-|-|
| standard keys (NEC1 signal) | NEC                    |                    NEC |
| arrow keys (NEC2 signal) | NEC | NEC2 | 

### Parametric Remote

| Input  | Repeat Finder off | Repeat Finder on |
|-|-|-|
| standard keys (NEC1 signal) | NEC1 | NEC |
| arrow keys (NEC2 signal) | NEC2 | NEC2 |

**Comment by bengtmartensson** [comment]
Sorry for that. Please try the new snapshot.

Note that the repeat finder will _by necessity_ squash a NEC2-sequence into NEC.

**Reply by eriknz** [reply]
Thanks. I'm still getting a couple of unexpected results, namely in Parametric Remote with Repeat Finder on, NEC1 is decoded as NEC, and in the Scrutinizer with Repeat Finder on, NEC2 is decoded as NEC2 even though you say it shouldn't be. This time I'm running the .AppImage in Debian as there was no exe, not sure if that makes a difference. Anyway, my reason for trying IrScrutinizer was to decode a few remotes, which I can do now so don't feel there's any rush to update it. I can do more tests if it'll help. Thanks.

### Scrutinize signal panel, continuous capture.

| Input  | Repeat Finder off | Repeat Finder on |
|-|-|-|
| standard keys (NEC1 signal) | NEC1 | NEC1 |
| arrow keys (NEC2 signal) | NEC2 | NEC2 | 

### Parametric Remote

| Input  | Repeat Finder off | Repeat Finder on |
|-|-|-|
| standard keys (NEC1 signal) | NEC1 | NEC |
| arrow keys (NEC2 signal) | NEC2 | NEC2 |

**Reply by bengtmartensson** [reply]
> in Parametric Remote with Repeat Finder on, NEC1 is decoded as NEC

Cannot reproduce; I get the correct result (unless "super short" press).

> in the Scrutinizer with Repeat Finder on, NEC2 is decoded as NEC2 even though you say it shouldn't be. 

OK. I see that the statement was not correct, so much better. It is true for the official version though... ;-).

So I think we can consider this as solved. Again, thank you for bringing this problem to my attention.



> as there was no exe

The ci-build build process does not build the exe, so I have to upload that manually. I do this sometimes, sometimes not. And always if some explicitly requests it...
